Server doesn't shut down cleanly after service failure during startup

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
If starting up the `JerseyService` fails, e.g. due to a misconfigured port, the server should initiate a shutdown and the process should terminate.

## Current Behavior
If starting up a the `JerseyService` fails, the server initiates a shutdown. However, it then hangs and the process has to be killed manually.

## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->
In the stack trace below, the only service mentioned to be not `TERMINATED` is the JobSchedulerService. Looking at a thread dump of the process reveals that we are getting stuck waiting for the server to reach the `RUNNING` state [here](https://github.com/Graylog2/graylog2-server/blob/9a47813f0b9bbe3bbacf545dba4e29ffa8982080/graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java#L78).
So as the server never reaches that state, we will have to find a way to make the job scheduler shut down nevertheless.

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. Change you `graylog.conf` by setting your `http_bind_address` to `http_bind_address = 127.0.0.1:1`
2. Start the Graylog Server
3. Watch the logs to see it shut down after failing to start up correctly

## Context
Stacktrace:
<pre>
2020-08-10 14:13:16,782 ERROR: org.graylog2.bootstrap.ServerBootstrap - Unable to shutdown properly on time. {STOPPING=[JobSchedulerService [STOPPING]], TERMINATED=[InputSetupService [TERMINATED], MongoDBProcessingStatusRecorderService [TERMINATED], UrlWhitelistService [TERMINATED], StreamCacheService [TERMINATED], OutputSetupService [TERMINATED], LookupTableService [TERMINATED], PeriodicalsService [TERMINATED], EtagService [TERMINATED], ConfigurationEtagService [TERMINATED], GracefulShutdownService [TERMINATED], JournalReader [TERMINATED], KafkaJournal [TERMINATED], BufferSynchronizerService [TERMINATED]], FAILED=[JerseyService [FAILED]]}
2020-08-10 14:13:16,782 ERROR: org.graylog2.bootstrap.ServerBootstrap - Graylog startup failed. Exiting. Exception was:
java.lang.IllegalStateException: Expected to be healthy after starting. The following services are not running: {FAILED=[JerseyService [FAILED]]}
	at com.google.common.util.concurrent.ServiceManager$ServiceManagerState.checkHealthy(ServiceManager.java:773) ~[guava-28.2-jre.jar:?]
	at com.google.common.util.concurrent.ServiceManager$ServiceManagerState.awaitHealthy(ServiceManager.java:585) ~[guava-28.2-jre.jar:?]
	at com.google.common.util.concurrent.ServiceManager.awaitHealthy(ServiceManager.java:316) ~[guava-28.2-jre.jar:?]
	at org.graylog2.bootstrap.ServerBootstrap.startCommand(ServerBootstrap.java:161) [classes/:?]
	at org.graylog2.bootstrap.CmdLineTool.run(CmdLineTool.java:210) [classes/:?]
	at org.graylog2.bootstrap.Main.main(Main.java:50) [classes/:?]
	Suppressed: com.google.common.util.concurrent.ServiceManager$FailedService: JerseyService [FAILED]
	Caused by: java.net.SocketException: Permission denied
		at sun.nio.ch.Net.bind0(Native Method) ~[?:1.8.0_222]
		at sun.nio.ch.Net.bind(Net.java:433) ~[?:1.8.0_222]
		at sun.nio.ch.Net.bind(Net.java:425) ~[?:1.8.0_222]
		at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223) ~[?:1.8.0_222]
		at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74) ~[?:1.8.0_222]
		at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bindToChannelAndAddress(TCPNIOBindingHandler.java:107) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOBindingHandler.bind(TCPNIOBindingHandler.java:64) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:215) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:195) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.nio.transport.TCPNIOTransport.bind(TCPNIOTransport.java:186) ~[grizzly-framework-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.http.server.NetworkListener.start(NetworkListener.java:711) ~[grizzly-http-server-2.4.4.jar:2.4.4]
		at org.glassfish.grizzly.http.server.HttpServer.start(HttpServer.java:256) ~[grizzly-http-server-2.4.4.jar:2.4.4]
		at org.graylog2.shared.initializers.JerseyService.startUpApi(JerseyService.java:201) ~[classes/:?]
		at org.graylog2.shared.initializers.JerseyService.startUp(JerseyService.java:147) ~[classes/:?]
		at com.google.common.util.concurrent.AbstractIdleService$DelegateService$1.run(AbstractIdleService.java:62) ~[guava-28.2-jre.jar:?]
		at com.google.common.util.concurrent.Callables$4.run(Callables.java:119) ~[guava-28.2-jre.jar:?]
		at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_222]
</pre>

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

* Graylog Version: `master`
* Java Version:
* Elasticsearch Version:
* MongoDB Version:
* Operating System:
* Browser version:

--- a/graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java
+++ b/graylog2-server/src/main/java/org/graylog/scheduler/JobSchedulerService.java
@@ -48,6 +48,7 @@ public class JobSchedulerService extends AbstractExecutionThreadService {
     private final CountDownLatch shutdownLatch = new CountDownLatch(1);
     private final Duration loopSleepDuration;
     private final InterruptibleSleeper sleeper = new InterruptibleSleeper();
+    private Thread executionThread;
 
     @Inject
     public JobSchedulerService(JobExecutionEngine.Factory engineFactory,
@@ -69,14 +70,21 @@ public class JobSchedulerService extends AbstractExecutionThreadService {
     @Override
     protected void startUp() throws Exception {
         schedulerEventBus.register(this);
+        this.executionThread = Thread.currentThread();
     }
 
     @Override
     protected void run() throws Exception {
         // Safety measure to make sure everything is started before we start job scheduling.
         LOG.debug("Waiting for server to enter RUNNING status before starting the scheduler loop");
-        serverStatus.awaitRunning(() -> LOG.debug("Server entered RUNNING state, starting scheduler loop"));
-
+        try {
+            serverStatus.awaitRunning();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.debug("Was interrupted while waiting for server to enter RUNNING state. Aborting.");
+            return;
+        }
+        LOG.debug("Server entered RUNNING state, starting scheduler loop");
 
         if (schedulerConfig.canStart()) {
             boolean executionEnabled = true;
@@ -124,6 +132,7 @@ public class JobSchedulerService extends AbstractExecutionThreadService {
         schedulerEventBus.unregister(this);
         shutdownLatch.countDown();
         jobExecutionEngine.shutdown();
+        executionThread.interrupt();
     }
 
     /**

--- a/graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java
+++ b/graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java
@@ -153,7 +153,9 @@ public class ServerStatus {
      * Blocks until the server enters the RUNNING state and then executes the given Runnable.
      * <p>
      * <b>This method is not interruptible while waiting for the server to enter the RUNNING state.</b>
+     * @deprecated Preferably use {@link #awaitRunning()} instead, which is interruptible.
      */
+    @Deprecated
     public void awaitRunning(final Runnable runnable) {
         LOG.debug("Waiting for server to enter RUNNING state");
         Uninterruptibles.awaitUninterruptibly(runningLatch);

--- a/graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java
+++ b/graylog2-server/src/main/java/org/graylog2/plugin/ServerStatus.java
@@ -149,6 +149,11 @@ public class ServerStatus {
         publishLifecycle(Lifecycle.OVERRIDE_LB_THROTTLED);
     }
 
+    /**
+     * Blocks until the server enters the RUNNING state and then executes the given Runnable.
+     * <p>
+     * <b>This method is not interruptible while waiting for the server to enter the RUNNING state.</b>
+     */
     public void awaitRunning(final Runnable runnable) {
         LOG.debug("Waiting for server to enter RUNNING state");
         Uninterruptibles.awaitUninterruptibly(runningLatch);
@@ -160,6 +165,15 @@ public class ServerStatus {
         } catch (Exception e) {
             LOG.error("awaitRunning callback failed", e);
         }
+    }
+
+    /**
+     * Blocks until the server enters the RUNNING state.
+     * @throws InterruptedException if the thread is interrupted while waiting for the server to enter the RUNNING
+     * state.
+     */
+    public void awaitRunning() throws InterruptedException {
+        runningLatch.await();
     }
 
     public DateTime getStartedAt() {

